### DISCUSSION TITLE
Expose data via JSON

### DISCUSSION THREAD

**Discussion by mattstauffer** [laravel/valet#1138] [Ideas]
In our [stream yesterday](https://www.youtube.com/watch?v=uTb-xCrvcDg) @mpociot & I were talking about the idea of exposing some Valet configuration and details (e.g. linked and parked sites) via JSON. 

We ended up instead requiring Valet as Composer dependency, but today I talked to @nicoverbruggen who built [PHPmon](https://github.com/nicoverbruggen/phpmon) and he was saying it would be valuable to his tool, and probably others, to have this information available as JSON.

Some ideas of data we could expose:

## High level config

- Config object, which reads all config items (which right now is *just* `~/config/valet/config.json` but could later be more)
- Possible smarter keys for any required options, which have internal defaults
    - TLD (assuming it really does default to `.test` if it's missing from config)
    - Directory listing

## Parked and linked sites

- Single flat list of all sites, each of which would give:
    - Driver
    - Path
    - Array of Links
       - Is this parked or linked
       - Domain
       - Whether it's secured
   - Wish list
       - PHP constraints
       - Laravel version
       - Checkmate-style stuff (e.g. is it out of date on Laravel?)

**Comment by mattstauffer** [comment]
Thinking

```
valet info --json
```

MAY just give everything in one fell swoop?

We could add JSON output to `links` and `parked` and `paths` but that doesn't expose the config and doesn't give us the nice flattened version either.

**Comment by drbyte** [comment]
Some of those operations are more expensive to run than others, so it may be valuable to allow being specific about which things one wants to query, perhaps with additional `--foo` params.

**Comment by nicoverbruggen** [comment]
How about having

```
valet --json
```

return the info about the current Valet installation (version number and configuration?). 

I forgot to mention that adding the version number to this JSON object would be helpful.

It could look like this:

```json
{
    "version": "2.13.16",
    "config": {
        "tld": "test",
        "paths": [
            "/Users/nicoverbruggen/.config/valet/Sites"
        ]
    }
}
```

The other commands, `valet links --json`, `valet parked --json` and `valet paths --json` can remain properly specific as you proposed.

**Comment by mattstauffer** [comment]
That's a great idea, although I still think I'd probably go for `valet info --json` for the simpler version you proposed, @nicoverbruggen.

Good thinking on the cost @drbyte!
